src/Security/Authentication/Negotiate/src/NegotiateHandler.cs

```diff
@@ -0,0 +1,334 @@
+        /// <param name="logger"></param>
+        /// <param name="encoder"></param>
+        /// <param name="clock"></param>
+        public NegotiateHandler(IOptionsMonitor<NegotiateOptions> options, ILoggerFactory logger, UrlEncoder encoder, ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        { }
+
+        /// <summary>
+        /// The handler calls methods on the events which give the application control at certain points where processing is occurring. 
+        /// If it is not provided a default instance is supplied which does nothing when the methods are called.
+        /// </summary>
+        protected new NegotiateEvents Events
+        {
+            get => (NegotiateEvents)base.Events;
+            set => base.Events = value;
+        }
+
+        /// <summary>
+        /// Creates the default events type.
+        /// </summary>
+        /// <returns></returns>
+        protected override Task<object> CreateEventsAsync() => Task.FromResult<object>(new NegotiateEvents());
+
+        private bool IsHttp2 => string.Equals("HTTP/2", Request.Protocol, StringComparison.OrdinalIgnoreCase);
+
+        /// <summary>
+        /// Intercepts incomplete Negotiate authentication handshakes and continues or completes them.
+        /// </summary>
+        /// <returns>True if a response was generated, false otherwise.</returns>
+        public async Task<bool> HandleRequestAsync()
+        {
+            try
+            {
+                if (_requestProcessed)
+                {
+                    // This request was already processed but something is re-executing it like an exception handler.
+                    // Don't re-run because we could corrupt the connection state, e.g. if this was a stage2 NTLM request
+                    // that we've already completed the handshake for.
+                    return false;
+                }
+
+                _requestProcessed = true;
+
+                if (IsHttp2)
+                {
+                    // HTTP/2 is not supported. Do not throw because this may be running on a server that supports
+                    // both HTTP/1 and HTTP/2.
+                    return false;
+                }
+
+                var connectionItems = GetConnectionItems();
+                var persistence = (AuthPersistence)connectionItems[AuthPersistenceKey];
+                _negotiateState = persistence?.State;
+
+                var authorizationHeader = Request.Headers[HeaderNames.Authorization];
+
+                if (StringValues.IsNullOrEmpty(authorizationHeader))
+                {
+                    if (_negotiateState?.IsCompleted == false)
+                    {
+                        throw new InvalidOperationException("An anonymous request was received in between authentication handshake requests.");
+                    }
+                    Logger.LogDebug($"No Authorization header");
+                    return false;
```

Comment:
Does prefixing with a $ add any extra runtime overhead when you don't actually have any dynamic content in the string?

Reply:
Leftover from when I had more parameters. Removed.

Reply:
To answer your original question (since it seemed interesting): No, the compiler optimizes it out:

https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgBgATIIwG4CwAUMgMyYBM6AwugN6HoOanIAs6AcgPYAKAhgE68AtgGcAFAEpa9RrOwBOMQBIARADFOnFRPwFZAXxkMjTTGwDqYYAAs+g0ZOl7ZjBcvWbaLMvu26DhPpAA=